habitat_baselines/models/resnet_encoders.py

```diff
@@ -226,21 +283,47 @@ def forward(self, observations):
 
         def resnet_forward(observation):
             resnet_output = torch.zeros(
-                observation.size(0), self.resnet_layer_size
-            ).to(self.device)
+                1, dtype=torch.float32, device=self.device
+            )
 
             def hook(m, i, o):
-                # self.resnet_output.resize_(d.size())
-                resnet_output.copy_(torch.flatten(o, 1).data)
+                if self.spatial_output:
+                    resnet_output.set_(o)
+                else:
+                    resnet_output.set_(torch.flatten(o, 1).data)
 
             # output: [BATCH x RESNET_DIM]
             h = self.layer_extract.register_forward_hook(hook)
             self.cnn(observation)
             h.remove()
             return resnet_output
 
-        # permute tensor to dimension [BATCH x CHANNEL x HEIGHT x WIDTH]
-        rgb_observations = observations["rgb"].permute(0, 3, 1, 2)
-        rgb_observations = rgb_observations / 255.0  # normalize RGB
-        resnet_output = resnet_forward(rgb_observations)
-        return self.activation(self.fc(resnet_output))  # [BATCH x OUTPUT_DIM]
+        if "rgb_features" in observations:
+            resnet_output = observations["rgb_features"]
+        else:
+            # permute tensor to dimension [BATCH x CHANNEL x HEIGHT x WIDTH]
+            rgb_observations = observations["rgb"].permute(0, 3, 1, 2)
+            rgb_observations = rgb_observations / 255.0  # normalize RGB
+            resnet_output = resnet_forward(rgb_observations)
+
+        if self.spatial_output:
+            b, c, h, w = resnet_output.size()
+
+            spatial_features = (
+                self.spatial_embeddings(
+                    torch.arange(
+                        0,
+                        self.spatial_embeddings.num_embeddings,
+                        device=resnet_output.device,
+                        dtype=torch.long,
+                    )
+                )
+                .view(1, -1, h, w)
+                .expand(b, self.spatial_embeddings.embedding_dim, h, w)
+            )
+
+            return torch.cat([resnet_output, spatial_features], dim=1)
+        else:
+            return self.activation(
+                self.fc(resnet_output)
+            )  # [BATCH x OUTPUT_DIM]
```

Comment:
This linear layer expects resnet_output to be `[BATCH x linear_layer_input_size]` but when I ran it with VLNBaselinePolicy, resnet_output was sometimes `[BATCH x linear_layer_input_size x 1 x 1]` and it errored out. We could maybe squeeze and check if batch if batch size is 1:

            resnet_output = resnet_output.squeeze()
            if len(resnet_output.size()) == 1:
                resnet_output = resnet_output.unsqueeze(0)

Both VLNBaselinePolicy and VLNRCMPolicy ran on my machine once this was in place.

Reply:
Ah, so it happens when caching features.  The cached features don't get the squeeze.  I removed the flatten from the hook and added a call to it before the FC layer.  That felt cleaner than squeezing the features to cache for some reason :)